### PR TITLE
Add fixed issues to regex fuzzer

### PR DESCRIPTION
**Pull request by anthony-chang**
Closes https://github.com/NVIDIA/spark-rapids/issues/4603

This PR adds regex `|` choice operator, octal chars in character classes, and escaped characters to character class ranges to the fuzzer, and fixes some corner cases from these updates:
1.  When a hex or octal representation of a meta character is used, eg. `\x24` for the character `$`, Java treats it like `\$` but cuDF treats it as the end of line anchor character, so we need to escape these characters in the transpilation. 
2.  There was a bug when parsing octal chars in character classes where we would sometimes read 1 character past the octal digit
3. Zero-repetitions (ie, `{0}` and `{0,0}`) inside a capture group has some inconsistences so I have disabled them.

### CODE REVIEW

**Comment by anthony-chang** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala [hunk 1]
```diff
@@ -1442,11 +1446,28 @@ class CudfRegexTranspiler(mode: RegexMode) {
...
-              case true => throw new RegexUnsupportedException(
-                "Line and string anchors are not supported in capture groups", part.position)
-              case false =>
-            }}
+            parts.foreach { part => 
+              isBeginOrEndLineAnchor(part) match {
+                case true => throw new RegexUnsupportedException(
+                  "Line and string anchors are not supported in capture groups", part.position)
+                case false =>
+              }
```
Comment (andygrove) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:1454]:
```suggestion
              if (isBeginOrEndLineAnchor(part)) {
                throw new RegexUnsupportedException(
                  "Line and string anchors are not supported in capture groups", part.position)
              }

```

**Comment by andygrove** [comment]
LGTM. Left one code style suggestion.

**Review by andygrove — APPROVED** [review]

**Comment by anthony-chang** [comment]
build

**Review by andygrove — APPROVED** [review]

**Review by NVnavkumar — APPROVED** [review]

**Pull request merged**